Add Cargo-style project discovery for Buck and Bazel Users

davidbarsky opened the pull request:
This feature requires the user to add a command that generates a `rust-project.json` from a set of files. Project discovery can be invoked in two ways:

1. At extension activation time, which includes the generated `rust-project.json` as part of the linkedProjects argument in `InitializeParams`.
2. Through a new command titled "rust-analyzer: Add current file to workspace", which makes use of a new, rust-analyzer-specific LSP request that adds the workspace without erasing any existing workspaces. Note that there is no mechanism to _remove_ workspaces other than "quit the rust-analyzer server".

Few notes:
-  I think that the command-running functionality _could_ merit being placed into its own extension (and expose it via extension contribution points) to provide build-system idiomatic progress reporting and status handling, but I haven't (yet) made an extension that does this nor does Buck expose this sort of functionality.
-  This approach would _just work_ for Bazel. I'll try and get the tool that's responsible for Buck integration open-sourced soon.
- On the testing side of things, I've used this in around my employer's Buck-powered monorepo and it's a nice experience. That being said, I can't think of an open-source repository where this can be tested in public, so you might need to trust me on this one.

I'd love to get feedback on:
- Naming of LSP extensions/new commands. I'm not too pleased with how "rust-analyzer: Add current file to workspace" is named, in that it's creating a _new_ workspace. I think that this command being added should be gated on `rust-analyzer.discoverProjectCommand` on being set, so I can add this in sequent commits.
- My Typescript. It's not particularly good.
- Suggestions on handling folders with _both_ Cargo and non-Cargo build systems and if I make activation a bit better.

(I previously tried to add this functionality entirely within rust-analyzer-the-LSP server itself, but matklad was right—an extension side approach is much, much easier.)

bjorn3 commented on crates/project-model/src/cfg_flag.rs:
```diff
@@ -38,6 +39,18 @@ impl<'de> serde::Deserialize<'de> for CfgFlag {
     }
 }
 
+impl Serialize for CfgFlag {
```
What are these serialization implementations used for? I couldn't find any serialization calls introduced in this PR.

davidbarsky replied:
`lsp-types-0.94` requires that [`Params` implement `DeserializeOwned + Serialize`](https://github.com/gluon-lang/lsp-types/blob/master/src/request.rs#L6), which feels unfortunate in this instance.

davidbarsky replied:
I've removed these implementations since I've moved to a "notify rust-analyzer that the configuration has changed"-based approach.

bjorn3 commented on editors/code/package.json:
```diff
@@ -447,6 +452,17 @@
                     ]
                 },
+                "rust-analyzer.discoverProjectCommand": {
+                    "markdownDescription": "Sets the command that rust-analyzer uses to generate `rust-project.json` files. This command is\n only suggested if a build system like Buck or Bazel is used. The command must accept files as arguements and return \n a rust-project.json over stdout.",
```
*arguments

davidbarsky replied:
whoops, thanks.

Veykril commented on crates/rust-analyzer/src/handlers.rs:
```diff
@@ -46,6 +47,22 @@ use crate::{
+    let config = Arc::make_mut(&mut state.config);
+    config.add_linked_projects(params.project);
+
     state.fetch_workspaces_queue.request_op("reload workspace request".to_string());
     state.fetch_build_data_queue.request_op("reload workspace request".to_string());
```
```suggestion
    state.fetch_workspaces_queue.request_op("linked projects changed".to_string());
    state.fetch_build_data_queue.request_op("linked projects changed".to_string());
```

Veykril commented on editors/code/package.json:
```diff
@@ -199,6 +199,11 @@
             },
+            {
+                "command": "rust-analyzer.addProject",
+                "title": "Add current file to workspace",
```
Maybe something like `Add project of current file to workspace`? (quite annoying how workspace has so many different meanings with cargo and vscode involved

davidbarsky replied:
That can work. I'll think of some other options (if I can) and get back to you on this in a few hours.

davidbarsky replied:
i've renamed it to "Add current file's crate to workspace", which is a bit more precise.

Veykril commented on editors/code/package.json:
```diff
@@ -447,6 +452,17 @@
                         "Fill missing expressions with reasonable defaults, `new` or `default` constructors."
                     ]
                 },
+                "rust-analyzer.discoverProjectCommand": {
```
You need to put this config before the `$generated-start` key, otherwise the tests that regen this file will fail

davidbarsky replied:
ty, fixed!

Veykril commented:
I feel like an `addProject` extension isn't the right tool here, instead I think the client should track all linked projects (including those of virtual project-json's) itself and tell the server to reload with the fresh set of supplied projects on change of these projects.

davidbarsky commented:
> I feel like an `addProject` extension isn't the right tool here, instead I think the client should track all linked projects (including those of virtual `project-json`'s) itself and tell the server to reload with the fresh set of supplied projects on change of these _projects_.

I'm not opposed to this change. I think it'd allow for a Buck/Bazel extension to manage linked projects and (potentially) surface them through a UI and mapping each linked project to a BUILD/BUCK file. One question, though: `linkedProjects` only gets sent _once_, during the initial server setup as a part of `InitializeParams`. I don't think it's a good idea to _fully_ reinitialize r-a on each reload, but I can change `rust-analyzer/reload` to accept `linkedProjects` (or even a full `ConfigData`!) instead of just `()`. Is that approach okay with you?

Veykril commented:
Yes, I forgot to add that, we'd still need this to be a request, repurposing `reloadWorkspace` sounds good.

> or even a full ConfigData

LSP already has a configuration changed notification so we could also re-use this and just update the linked projects key there. The main problem here is figuring out how to shove an "inline" rust project specification into there since right not this takes a list of file paths only

davidbarsky commented:
I've added a workspace configuration-based approach to reloading the workspace in cef2d6219af32e11e5713d14391d4086bb1b1602. I haven't removed the `rust-analyzer/AddProject`-based approach _yet_ because the way I've implementing reloading feels like an abuse of `vscode.WorkspaceConfiguration`.

davidbarsky commented:
I think I've addressed all the comments:
- I've rebased against the master branch.
- I've removed `rust-analyzer/addProject` in favor of notifying the server that configuration has changed. Note that the `rust-project.json` is never written to disk; it's synthesized entirely in-memory and placed into the `InitializeParams`. Actually updating VS Code's configuration _would_ result in writing to disk into the workspace-level `.vscode/` folder, but I think exposing this detail to users is a poor UX.

Veykril commented:
Will take a proper look tomorrow but

> Actually updating VS Code's configuration would result in writing to disk into the workspace-level .vscode/ folder, but I think exposing this detail to users is a poor UX.

fully agree with this, we should not be changing the user config. We only need to synthesize the `linkedProjects` field when passing it to the server, this way we could in the future also expose an API for the vscode extension that other extensions can hook into (which I think is what matklad was describing in the other issue at some point)

davidbarsky commented:
> Will take a proper look tomorrow

No worries, take care!

> fully agree with this, we should not be changing the user config. We only need to synthesize the `linkedProjects` field when passing it to the server, this way we could in the future also expose an API for the vscode extension that other extensions can hook into

Yup, that's the approach that I took here. It seems that writing out the config was part of VS Code-specific behavior when a key was updated in `vscode.WorkspaceConfiguration`.

> which I think is what matklad was describing in the other issue at some point

He was. I tried a different approach, and when that I didn't work out, I was able to get _this_ approach working in about 2 days. I'm a bit annoyed with myself that I didn't listen to him first!

Veykril commented on editors/code/src/config.ts:
```diff
@@ -280,18 +287,24 @@ export class Config {
-export function substituteVSCodeVariables<T>(resp: T): T {
+export function prepareVSCodeConfig<T>(
+    resp: T,
+    cb?: (key: Extract<keyof T, string>, res: { [key: string]: any }) => void
```
comment as to what this callback is for would be nice

davidbarsky replied:
documented this!

(I'm not sure I'm happy with this approach overall, but it seems to work well for now...)

Veykril commented on editors/code/src/client.ts:
```diff
@@ -95,7 +95,9 @@ export async function createClient(
                     if (resp && Array.isArray(resp)) {
                         return resp.map((val) => {
-                            return substituteVSCodeVariables(val);
+                            return prepareVSCodeConfig(val, (key, cfg) => {
+                                cfg[key] = config.discoveredWorkspaces;
+                            });
```
I don't quite follow this but doesn't this overwrite everything in the config?

davidbarsky replied:
whoops, I think you're right about that. Thanks for the catch!

Veykril commented on editors/code/src/ctx.ts:
```diff
@@ -169,7 +188,28 @@ export class Ctx {
+                rawInitializationOptions,
+                (key, obj) => {
+                    if (key === "linkedProjects") {
+                        obj["linkedProjects"] = this.config.discoveredWorkspaces;
```
same here

Veykril reviewed (changes requested):
One more thing

Veykril commented:
I am not sure whether this is the right way we want to tackle this (that is having this command be part of r-a in long term), but for the time being I'd say that's okay.

davidbarsky commented on editors/code/src/client.ts:
```diff
@@ -95,7 +95,11 @@ export async function createClient(
-                            return substituteVSCodeVariables(val);
+                            return prepareVSCodeConfig(val, (key, cfg) => {
+                                if (key === "linkedProjects") {
+                                    cfg[key] = config.discoveredWorkspaces;
```
`config.discoveredWorkspaces` is never null/undefiend, but it might be empty. at the moment, this should probably be something like `if (key === linkedProjects && !config.discoveredWorkspaces.isEmpty) { }`, but if you'd like, I can make the definition of `config.discoveredWorkspaces` `JsonProject[] | undefined`?

Veykril replied:
Right an empty check is better here then

davidbarsky replied:
will make that change, then!

davidbarsky replied:
alright, should be fixed. sorry about the back-and-forth on this.

davidbarsky commented:
> I am not sure whether this is the right way we want to tackle this (that is having this command be part of r-a in long term), but for the time being I'd say that's okay.

My feeling is that I'd like to get some user feedback first, but I strongly suspect that this should live in an external extension that rust-analyzer coordinates with. I don't expect rust-analyzer to worry about maintaining this for an extended period of time.

davidbarsky commented:
Sorry about the back and forth, but thank you for your patience :)

Veykril commented:
@bors r+

bors commented:
:pushpin: Commit 6e7bc07cdf225aea811f793c2f712f25846b8d20 has been approved by `Veykril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=Veykril 6e7bc07cdf225aea811f793c2f712f25846b8d20 -->
<!-- homu: {"type":"Approved","sha":"6e7bc07cdf225aea811f793c2f712f25846b8d20","approver":"Veykril","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:hourglass: Testing commit 6e7bc07cdf225aea811f793c2f712f25846b8d20 with merge c15335c8b02dbc0ccc04f04c56d6471120db1bd2...
<!-- homu: {"type":"BuildStarted","head_sha":"6e7bc07cdf225aea811f793c2f712f25846b8d20","merge_sha":"c15335c8b02dbc0ccc04f04c56d6471120db1bd2"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/4418708619/jobs/7746405766)
Approved by: Veykril
Pushing c15335c8b02dbc0ccc04f04c56d6471120db1bd2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Veykril","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/4418708619/jobs/7746405766"},"merge_sha":"c15335c8b02dbc0ccc04f04c56d6471120db1bd2"} -->

woody77 commented:
When is the discovery command run?  Is it run in lieu of opening a `rust-project.json` file?

davidbarsky commented:
> When is the discovery command run? Is it run in lieu of opening a rust-project.json file?

@woody77 At the moment, it can run in two situations:
- when the extension is activating (from either an editor restart or a Rust file being opened for the first time), or
- when the user runs the "Add Project to Current Workspace" command through the command palette.

To be entirely honest, I think I was a bit too conservative with the latter option. I think that we should be probably be running the discovery command each time the user opens a Rust file that isn't part of the current workspace—rust-analyzer is *stupidly* fast at loading projects, and getting the results from Buck takes a few seconds _at worst_.
